File: leetcode/leetcode_py/unique-number-of-occurrences.py
'''
https://leetcode.com/problems/unique-number-of-occurrences

1. Using map and set
time: o(n)
space: o(n)
'''
class Solution:
    def uniqueOccurrences(self, arr: List[int]) -> bool:
        m = defaultdict(int)
        s = set()
        for num in arr:
            m[num] += 1
        for key in m:
            if m[key] in s: return False
            s.add(m[key])
        return True
        

File: leetcode/leetcode_cpp/unique-number-of-occurrences.cpp
/*
https://leetcode.com/problems/unique-number-of-occurrences

1. Using map and set
time: o(n)
space: o(n)
*/
class Solution {
public:
    bool uniqueOccurrences(vector<int>& arr) {
        unordered_map<int, int> m;
        unordered_set<int> s;
        for (auto value : arr) m[value] += 1;
        for (auto& values : m) {
            //if (s.count(values.second)) return false;
            //s.insert(values.second);
            if (!s.insert(values.second).second) return false;
        }
        return true;
    }
};